File: javascript/easy/intersectionOfTwoArraysII.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function(nums1, nums2) {
    const map = new Map();
    let res = [];
    for(let num of nums1) {
        let count = map.get(num);
        map.set(num, (count === undefined) ? 1 : count+1);
    }
    for(let num of nums2) {
        let count = map.get(num);
        if(count !== undefined) {
            res.push(num);
            if(count == 1) {
                map.delete(num);
            } else {
                map.set(num, count-1);
            }
        }
    }
    return res;
};

File: ruby/easy/intersection_of_two_arrays_ii.rb
# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Integer[]}
def intersect(nums1, nums2)
    map,res = {}, []
    nums1.each do |num|
        map[num] = (map[num] ? map[num] : 0) + 1
    end
    nums2.each do |num|
        res.push(num) and map[num] -= 1 if map[num]
        map[num] = nil if map[num] == 0
    end
    return res
end
